feat(global, login)

holeman79 opened the pull request:
토이 프로젝트에서 진행했던 global이랑 login 기능

holeman79 committed:
feat(global, login)

- 초기 셋팅
- login 기능 추가

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/global/config/AuditingConfig.kt:
```diff
@@ -0,0 +1,18 @@
+    @Bean
+    fun auditorProvider(): AuditorAware<String>? {
+        //        Spring Security 적용 후 주석 된 코드로 대체
+        //        return () -> Optional.of(SecurityContextHolder.getContext().getAuthentication().getPrincipal().toString());
```
@holeman79 
원진님, 해당 코드는 현재 주석 처리된 내용으로 대체 되어야 할 것 같은데 한 번 확인 부탁드립니다!

holeman79 replied:
아 그렇네요 대체하도록 할게요!

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/global/domain/security/JwtProvider.kt:
```diff
@@ -0,0 +1,63 @@
+import java.security.Key
+import java.time.LocalDateTime
+import java.time.ZoneId
+import java.util.*
```
@holeman79 
Ktlint 에서는 와일드 카드 사용을 권장하지 않고 있습니다. ([참고 아티클](https://velog.io/@kshired/Ktlint%EB%8A%94-%EC%99%9C-wild-card-import%EB%A5%BC-%EA%B6%8C%EC%9E%A5%ED%95%98%EC%A7%80-%EC%95%8A%EC%9D%84%EA%B9%8C))
해당 부분은 인텔리제이에서 설정 가능하니 한 번 확인 부탁드립니다! ([설정 방법](https://blog.leocat.kr/notes/2020/12/14/intellij-avoid-wildcard-imports-in-kotlin-with-intellij))

holeman79 replied:
넵 해당 부분 반영하도록 하겠습니다

holeman79 replied:
마지막 행 개행은 알려주신대로 설정은 했는데 전체 소스에 적용하는 방법은 어떻게 하는지 모르겠네요....

seongheonsim replied:
아마 한 번 파일을 열어야 개행을 추가할 거에요ㅠㅠ

holeman79 replied:
파일 다 열어서 코드 마지막에 엔터치고 저장을 해야 반영이 되네요...ㅎㅎ 일단 코드가 얼마 안되니 다 적용해서 다시 커밋 올릴게요!

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/global/response/PageDto.kt:
```diff
@@ -0,0 +1,21 @@
+
+import org.springframework.data.domain.Page
+
+class PageDto<T>(val page: Page<T>) {
```
@holeman79 
`page` 프로퍼티가 응답 모델에 노출되어야 하는게 아니라면 `page: Page<T>` 같이 단순 입력 인자로 설정하는게 좋을 것 같습니다.

holeman79 replied:
오... 코틀린은 저걸 선언에서 뺄 수가 있군요ㅎㅎ 좋은 팁 감사합니다

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/global/response/PageDto.kt:
@holeman79 
해당 응답 모델의 프로퍼티가 private 접근 제어자로 설정되어 있으면 응답 시 노출이 안되는 것으로 알고 있어 해당 제어자는 제거해야 할 것 같아요! 
그리고 PageDto 클래스는 새로운 인스턴스가 생성된 후 데이터가 변경될 일이 없을 거 같아 모든 프로퍼티의 변수 선업 타입이 `val` 로 설정 되는 것이 좋을 것 같은데 어떻게 생각하실까요~?

holeman79 replied:
아 PageDto는 안바꿨었네요 변경하도록 하겠습니다. 그리고 추가적으로 PageDto도 이름 통일을 위해 PageResponse로 변경하도록 할게요

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/application/LoginService.kt:
```diff
@@ -0,0 +1,25 @@
+) {
+    fun login(loginCommand: LoginCommand): LoginResponse {
+        // 아이디가 존재하는지
+        val member = memberRepository.findByEmail(loginCommand.email) ?: throw MemberInputException("존재하지 않는 email입니다.")
```
@holeman79 
해당 예외를 던지면 응답 메세지에 아무런 내용이 노출되지 않는데 핸들러에 추가해야 하지 않을까 싶습니다.

holeman79 replied:
넵 GlobalExceptionHandler 에 추가해놓겠습니다

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/application/SignUpService.kt:
```diff
@@ -0,0 +1,23 @@
+    fun signUp(signUpCommand: SignUpCommand): SignUpResponse {
+        // 이미 존재하는 회원인지
+        if (memberRepository.existsByEmail(signUpCommand.email)) {
+            throw MemberInputException("이미 가입된 email 주소입니다.")
```
@holeman79 
해당 부분도 위의 리뷰와 동일합니다!

seongheonsim commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/domain/MemberRepository.kt:
```diff
@@ -0,0 +1,10 @@
+
+interface MemberRepository : JpaRepository<Member, Long> {
+
+    fun findByEmail(email: String): Member?
```
@holeman79 
`Optional<Member>` 로 리턴 타입을 사용하는건 어떠신가요?!

holeman79 replied:
Optional은 Java에서 Null을 처리하기 위해 필요한 클래스인데 코틀린은 safe call, elvis 연산자 등 null을 처리하기 위한 방법들이 존재해서 코틀린 라이브러리에는 Optional을 제공하지 않고 있습니다.(자바 라이브러리만 제공) [참고 아티클](https://medium.com/@limgyumin/%EC%BD%94%ED%8B%80%EB%A6%B0-nullable-%ED%83%80%EC%9E%85-vs-%EC%9E%90%EB%B0%94-optional-e698adc6d617) 참조 한 번 부탁드리겠습니다!

seongheonsim replied:
앗 실무에서는 `Optional` 로 받고 있어서 Nullable 로 리턴 받을 생각은 못해봤는데 말씀해주셔서 감사합니다!

seongheonsim reviewed (commented):
안녕하세요, 원진님!
저도 해당 부분 작업 중인데 제가 한 발 늦었네요!ㅎㅎ🏃‍♂️💨
몇 가지 리뷰 남겼으니 확인 부탁드립니다!

추가적으로 Ktlint 에서는 모든 파일의 마지막 행을 개행하는 것을 권장 하는데, 해당 부분도 인텔리제이에서 자동으로 추가해주는 방법이 있어 참고 내용 첨부 드립니다! ([참고 아티클](https://seongwon.dev/Git/20220303-%ED%8C%8C%EC%9D%BC%EC%9D%98_%EB%A7%88%EC%A7%80%EB%A7%89_%EA%B0%9C%ED%96%89/))

holeman79 committed:
refactor(global, login) : 리뷰 수정사항 반영

- AuditingConfig 주석 제거 및 SecurityContext Principal 사용자 적용
- 마지막 행 개행 적용
- import wildcard 제거
- MemberInputException 처리 핸들러 추가 및 BAD_MEMBER_INPUT 코드 추가
- PageDto -> PageResponse 이름 변경

seongheonsim reviewed (changes requested):
반영해주신 부분 확인 했습니다. 고생하셨습니다!
